sagemaker-neo-notebooks/edge/cpp-integration/tutorial.cc

```diff
@@ -0,0 +1,604 @@
+      "SampleAllocationTag", filename.c_str(), std::ios_base::in | std::ios_base::binary);
+  object_request.SetBody(input_data);
+
+  // put model to the s3 bucket
+  auto put_object_outcome = s3_client.PutObject(object_request);
+  if (!put_object_outcome.IsSuccess()) {
+    auto error = put_object_outcome.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("UploadModel error: " + error_str);
+  }
+}
+
+bool checkBucketExist(std::string& bucket_name) {
+  Aws::String s3_bucket_name(bucket_name.c_str(), bucket_name.size());
+  Aws::S3::S3Client s3_client = getS3Client();
+  auto list_bucket_resp = s3_client.ListBuckets();
+  if (list_bucket_resp.IsSuccess()) {
+    Aws::Vector<Aws::S3::Model::Bucket> bucket_list = list_bucket_resp.GetResult().GetBuckets();
+    for (auto const& bucket : bucket_list) {
+      if (bucket.GetName().compare(s3_bucket_name) == 0) {
+        return true;
+      }
+    }
+  } else {
+    auto error = list_bucket_resp.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("ListBucket error: " + error_str);
+  }
+  return false;
+}
+
+bool checkModelExist(std::string& bucket_name, std::string& model_name) {
+  Aws::String aws_bucket_name(bucket_name.c_str(), bucket_name.size());
+  Aws::String aws_model_name(model_name.c_str(), model_name.size());
+
+  Aws::S3::S3Client s3_client = getS3Client();
+  Aws::S3::Model::ListObjectsRequest objects_request;
+  objects_request.WithBucket(aws_bucket_name);
+
+  auto list_objects_outcome = s3_client.ListObjects(objects_request);
+  if (list_objects_outcome.IsSuccess()) {
+    Aws::Vector<Aws::S3::Model::Object> object_list =
+        list_objects_outcome.GetResult().GetContents();
+    for (auto const& s3_object : object_list) {
+      if (s3_object.GetKey().compare(aws_model_name) == 0) {
+        return true;
+      }
+    }
+  } else {
+    auto error = list_objects_outcome.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("ListObjects error: " + error_str);
+  }
+
+  return false;
+}
+
+void UploadModelToS3(std::string& model_name, std::string& filename, std::string& s3_bucket_name) {
+  // first create bucket
+  bool isBucketExist = checkBucketExist(s3_bucket_name);
+  if (!isBucketExist) {
+    createBucket(s3_bucket_name);
+  }
+
+  // then create/upload model
+  if (!checkModelExist(s3_bucket_name, model_name)) {
+    uploadModel(s3_bucket_name, model_name, filename);
+  }
+}
+
+nlohmann::json getIamPolicy() {
+  nlohmann::json statement;
+  statement["Action"] = "sts:AssumeRole";
+  statement["Effect"] = "Allow";
+  statement["Principal"]["Service"] = "sagemaker.amazonaws.com";
+
+  nlohmann::json policy;
+  policy["Statement"] = nlohmann::json::array();
+  policy["Statement"].push_back(statement);
+  policy["Version"] = "2012-10-17";
+  return policy;
+}
+
+bool checkIamRole(std::string& iam_role) {
```

Comment:
I probably don't make it very clear. Let's use const reference if the parameter is read only `const std::string& iam_role`. Otherwise, the API might break if someone wants to by const. Same to other places.

Reply:
revised as suggested